File: go/bccomponents/keypair.go
package bccomponents

import (
	bcrand "github.com/nickel-blockchaincommons/bcrand-go"
)

// Keypair generates a key pair using the default signature and encapsulation
// schemes (Schnorr for signing, X25519 for encapsulation).
func Keypair() (PrivateKeys, PublicKeys, error) {
	return KeypairOpt(SchemeSchnorr, EncapsulationX25519)
}

// KeypairUsing generates a key pair using the default schemes and a custom
// random number generator.
func KeypairUsing(rng bcrand.RandomNumberGenerator) (PrivateKeys, PublicKeys, error) {
	return KeypairOptUsing(SchemeSchnorr, EncapsulationX25519, rng)
}

// KeypairOpt generates a key pair with specified signature and encapsulation
// schemes.
func KeypairOpt(sigScheme SignatureScheme, encapScheme EncapsulationScheme) (PrivateKeys, PublicKeys, error) {
	sigPriv, sigPub, err := SigningKeypair(sigScheme)
	if err != nil {
		return PrivateKeys{}, PublicKeys{}, err
	}
	encapPriv, encapPub, err := EncapsulationKeypair(encapScheme)
	if err != nil {
		return PrivateKeys{}, PublicKeys{}, err
	}
	return NewPrivateKeys(sigPriv, encapPriv), NewPublicKeys(sigPub, encapPub), nil
}

// KeypairOptUsing generates a key pair with specified schemes using a custom
// random number generator.
func KeypairOptUsing(sigScheme SignatureScheme, encapScheme EncapsulationScheme, rng bcrand.RandomNumberGenerator) (PrivateKeys, PublicKeys, error) {
	sigPriv, sigPub, err := SigningKeypairUsing(sigScheme, rng, "")
	if err != nil {
		return PrivateKeys{}, PublicKeys{}, err
	}
	// EncapsulationKeypair does not currently accept an RNG parameter,
	// so we use the standard EncapsulationKeypair function.
	encapPriv, encapPub, err := EncapsulationKeypair(encapScheme)
	if err != nil {
		return PrivateKeys{}, PublicKeys{}, err
	}
	return NewPrivateKeys(sigPriv, encapPriv), NewPublicKeys(sigPub, encapPub), nil
}

File: csharp/BCComponents/BCComponents/Keypair.cs
using BlockchainCommons.BCRand;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// Static helper for generating complete keypairs.
/// </summary>
/// <remarks>
/// Provides convenience methods for generating complete signing and encapsulation
/// keypairs using Schnorr + X25519 as the default algorithms.
/// </remarks>
public static class Keypair
{
    /// <summary>
    /// Generates a new keypair using the system's cryptographically secure RNG.
    /// </summary>
    /// <remarks>
    /// Uses Schnorr for signing and X25519 for encapsulation by default.
    /// </remarks>
    /// <returns>A tuple of (<see cref="PrivateKeys"/>, <see cref="PublicKeys"/>).</returns>
    public static (PrivateKeys PrivateKeys, PublicKeys PublicKeys) Generate()
    {
        return GenerateUsing(SecureRandomNumberGenerator.Shared);
    }

    /// <summary>
    /// Generates a new keypair using the given random number generator.
    /// </summary>
    /// <remarks>
    /// Uses Schnorr for signing and X25519 for encapsulation by default.
    /// </remarks>
    /// <param name="rng">The random number generator to use.</param>
    /// <returns>A tuple of (<see cref="PrivateKeys"/>, <see cref="PublicKeys"/>).</returns>
    public static (PrivateKeys PrivateKeys, PublicKeys PublicKeys) GenerateUsing(IRandomNumberGenerator rng)
    {
        return GenerateUsing(rng, SignatureScheme.Schnorr, EncapsulationScheme.X25519);
    }

    /// <summary>
    /// Generates a new keypair using the given RNG and specific schemes.
    /// </summary>
    /// <param name="rng">The random number generator to use.</param>
    /// <param name="signingScheme">The signing scheme to use.</param>
    /// <param name="encapsulationScheme">The encapsulation scheme to use.</param>
    /// <returns>A tuple of (<see cref="PrivateKeys"/>, <see cref="PublicKeys"/>).</returns>
    public static (PrivateKeys PrivateKeys, PublicKeys PublicKeys) GenerateUsing(
        IRandomNumberGenerator rng,
        SignatureScheme signingScheme,
        EncapsulationScheme encapsulationScheme)
    {
        var (sigPriv, sigPub) = signingScheme.KeypairUsing(rng, "");
        var (encPriv, encPub) = encapsulationScheme.KeypairUsing(rng);

        var privateKeys = new PrivateKeys(sigPriv, encPriv);
        var publicKeys = new PublicKeys(sigPub, encPub);
        return (privateKeys, publicKeys);
    }
}
